### PR TITLE
Resolve #43

### LINKED ISSUE #43: Registering the service one should provide absolute path, otherwise exception is thrown
When registering service via `services.AddJsonLocalization`, if the path is not an absolute one it will throw the following exception when resolving the files by `IStringLocalizer`

```cs
// this will cause an error when _localizer["Hello"] is called later on
services.AddJsonLocalization(options => options.ResourcesPath = "Resources");
```

it should be changed to 

```cs
var directory = Path.Combine(Directory.GetCurrentDirectory(), "Resources");
services.AddJsonLocalization(options => options.ResourcesPath = directory);
```
the latter will work perfect.

I think that extra line can be moved to the library itself.

```bash
System.ArgumentException: The path must be absolute. (Parameter 'root')
   at Microsoft.Extensions.FileProviders.PhysicalFileProvider..ctor(String root, ExclusionFilters filters)
   at Microsoft.Extensions.FileProviders.PhysicalFileProvider..ctor(String root)
   at Microsoft.Extensions.Configuration.FileConfigurationExtensions.SetBasePath(IConfigurationBuilder builder, String basePath)
   at My.Extensions.Localization.Json.JsonStringLocalizer.<>c__DisplayClass17_0.<BuildResourcesCache>b__0(String _)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at My.Extensions.Localization.Json.JsonStringLocalizer.BuildResourcesCache(String culture)
   at My.Extensions.Localization.Json.JsonStringLocalizer.GetStringSafely(String name)
   at My.Extensions.Localization.Json.JsonStringLocalizer.get_Item(String name)
   at My.Extensions.Localization.Json.Internal.StringLocalizer.get_Item(String name)

HEADERS
=======
Accept-Language: fr-FR
```

### PR DESCRIPTION
**Pull request by shahabganji**
When building the configuration an absolute path should be provided.

### CODE REVIEW

File: src/My.Extensions.Localization.Json/JsonStringLocalizer.cs [hunk 1]
```diff
@@ -171,7 +171,7 @@ private void BuildResourcesCache(string culture)
                 if (File.Exists(_searchedLocation))
                 {
                     var builder = new ConfigurationBuilder()
-                    .SetBasePath(_resourcesPath)
+                    .SetBasePath(Path.Combine(Directory.GetCurrentDirectory(),_resourcesPath))
```
Comment (hishamco) [inline src/My.Extensions.Localization.Json/JsonStringLocalizer.cs:174]: We need to change this line instead

```csharp
_searchedLocation = Path.Combine(Directory.GetCurrentDirectory(), _resourcesPath, resourceFile);
```

File: samples/LocalizationSample/Controllers/TestController.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+using Microsoft.AspNetCore.Mvc;
```
Comment (hishamco) [inline samples/LocalizationSample/Controllers/TestController.cs:1]: Remove this class and replace it with a functional test

File: samples/LocalizationSample/Startup.cs [hunk 3]
```diff
@@ -14,6 +14,7 @@ public class Startup
     {
         public void ConfigureServices(IServiceCollection services)
         {
+            services.AddControllers();
```
Comment (hishamco) [inline samples/LocalizationSample/Startup.cs:17]: Revert the changes from this file

File: src/My.Extensions.Localization.Json/JsonStringLocalizer.cs (hunk 1 above)
Comment (shahabganji) [reply src/My.Extensions.Localization.Json/JsonStringLocalizer.cs]: the above-mentioned line has no effect on the issue, I changed it in the constructor

File: src/My.Extensions.Localization.Json/JsonStringLocalizer.cs [hunk 4]
```diff
@@ -25,7 +25,8 @@ public class JsonStringLocalizer : IStringLocalizer
             string resourceName, // TODO: Use optional parameter in upcoming major release
             ILogger logger)
         {
-            _resourcesPath = resourcesPath ?? throw new ArgumentNullException(nameof(resourcesPath));
+            var path = resourcesPath ?? throw new ArgumentNullException(nameof(resourcesPath));
```
Comment (hishamco) [inline src/My.Extensions.Localization.Json/JsonStringLocalizer.cs:28]: Why the below code doesn't work?

File: test/LocalizationSample.FunctionalTest/LocalizationSample.FunctionalTest.csproj [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        <PackageReference Include="FluentAssertions" Version="5.10.3" />
+        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
+        <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="3.1.2" />
+        <PackageReference Include="xunit" Version="2.4.0" />
+        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0" />
+        <PackageReference Include="coverlet.collector" Version="1.2.0" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\..\samples\LocalizationSample\LocalizationSample.csproj" />
```
Comment (hishamco) [inline test/LocalizationSample.FunctionalTest/LocalizationSample.FunctionalTest.csproj:19]: Can we do the test against `LocalizationSample.Mvc`

File: test/LocalizationSample.Mvc.FunctionalTest/LocalizationSample.Mvc.FunctionalTest.csproj [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        <PackageReference Include="FluentAssertions" Version="5.10.3" />
+        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
+        <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="3.1.2" />
+        <PackageReference Include="xunit" Version="2.4.0" />
+        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0" />
+        <PackageReference Include="coverlet.collector" Version="1.2.0" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\..\samples\LocalizationSample.Mvc\LocalizationSample.Mvc.csproj" />
```
Comment (hishamco) [inline test/LocalizationSample.Mvc.FunctionalTest/LocalizationSample.Mvc.FunctionalTest.csproj:19]: It's here

**Comment by hishamco** [comment]
I will add some commits to yours if you don't mind

File: test/LocalizationSample.Mvc.FunctionalTest/HomeControllerSpec.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            var response = await _client.GetAsync("/Home/Privacy").ConfigureAwait(false);
+            var result = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+            result.Should().Contain("Privacy Policy");
+        }
+
+        [Fact]
+        public async Task PrivacyPolicy_should_return_Politique_de_confidentialité()
+        {
+            var response = await _client.GetAsync("/Home/Privacy/?culture-fr-FR&ui-culture=fr-FR")
```
Comment (hishamco) [inline test/LocalizationSample.Mvc.FunctionalTest/HomeControllerSpec.cs:31]: `?culture-fr-FR` -> `?culture=fr-FR`

Comment (shahabganji) [reply test/LocalizationSample.Mvc.FunctionalTest/HomeControllerSpec.cs]: I surprised too, I saw that in the [image](https://github.com/hishamco/My.Extensions.Localization.Json/issues/43#issuecomment-608332510) you've sent me on the issue discussion. :+1:

File: test/LocalizationSample.FunctionalTest/LocalizationSample.FunctionalTest.csproj (hunk 5 above)
Comment (shahabganji) [reply test/LocalizationSample.FunctionalTest/LocalizationSample.FunctionalTest.csproj]: we did it on both, right?

File: src/My.Extensions.Localization.Json/JsonStringLocalizer.cs (hunk 4 above)
Comment (shahabganji) [reply src/My.Extensions.Localization.Json/JsonStringLocalizer.cs:28]: did not check the reason, just saw the tests are failing and when moving to are they are passing. I could investigate though.

**Comment by hishamco** [comment]
@shahabganji are you sure both of functional tests are passed?

**Comment by shahabganji** [comment]
@hishamco , yes pretty much sure 

Screen Shot 2020-04-03 at 17 23 54

**Comment by hishamco** [comment]
Strange!! It doesn't work with me ;)

FunctionalTest

File: src/My.Extensions.Localization.Json/JsonStringLocalizer.cs (hunk 4 above)
Comment (shahabganji) [reply src/My.Extensions.Localization.Json/JsonStringLocalizer.cs:28]: @hishamco because it changes the `_searchedLocation` it should change `_resourcesPath` on line [178](https://github.com/shahabganji/My.Extensions.Localization.Json/blob/hotfix/absolutepath/src/My.Extensions.Localization.Json/JsonStringLocalizer.cs#L178)

**Comment by hishamco** [comment]
Let me check what's prevent the test with French culture ...

**Comment by shahabganji** [comment]
Would you try `dotnet test` via the command line? 

Screen Shot 2020-04-03 at 17 33 11

**Comment by hishamco** [comment]
Sure

**Comment by hishamco** [comment]
Same thing, let me try something, coz the issue in the application root path

**Comment by hishamco** [comment]
Finally it got work!! I will update the PR soon, @shahabganji please check it again in your end

**Comment by shahabganji** [comment]
ping me whenever you've updated the PR, I will check it too.

**Comment by hishamco** [comment]
@shahabganji  as I said earlier the issue comes from the application root path, please check my notes in your PR

**Comment by hishamco** [comment]
There 're two failing tests I need to fix after this PR, please run your added functional tests and revert the unnecessary changes

I will merge this when you confirm

**Comment by shahabganji** [comment]
LGTM, all passed. well done 👍

**Comment by hishamco** [comment]
Thank you very much Saeed

**Comment by shahabganji** [comment]
My pleasure Hisham 👍

**Comment by hishamco** [comment]
FYI @shahabganji https://github.com/hishamco/My.Extensions.Localization.Json/releases/tag/v2.1, the NuGet package will be show up soon ...

**Comment by shahabganji** [comment]
Great! Thanks.
